NEW: Input consumption - Priorities instead of complexity [ISX-2510] (#2402)
Unity-Technologies/InputSystem: Packages/com.unity.inputsystem/Documentation~/Actions.md

You can't change certain aspects of the configuration, such as action bindings, while an action is enabled. To stop actions or action maps from responding to input, call  [`Disable`](xref:UnityEngine.InputSystem.InputAction.Disable).

While enabled, an action actively monitors the [controls](xref:input-system-controls) it's bound to. If a bound control changes state, the action processes the change. If the control's change represents an [interaction](xref:input-system-interactions) change, the action creates a response. All of this happens during the Input System update logic. Depending on the [update mode](xref:input-system-settings#update-mode) selected in the input settings, this happens once every frame, once every fixed update, or manually if updates are set to manual.

## Overlapping bindings and action priority

> When several enabled actions share the same physical control (for example a plain **B** key action and a **Shift+B** composite), the Input System can resolve which action should respond first. This is controlled in **Project Settings** > **Input System Package** under [Improved Shortcut Support](xref:input-system-settings#improved-shortcut-support).

Reviewer: ```suggestion
When several enabled actions share the same physical control (for example, a plain **B** key action and a **Shift**+**B** composite), the Input System can resolve which action should respond first. 

You can control this behavior in **Project Settings** > **Input System Package** > [**Improved Shortcut Support**](xref:input-system-settings#improved-shortcut-support).
```

| Setting | Behavior |
| ------- | -------- |

> | [Complexity-Based Shortcut Resolution](xref:UnityEngine.InputSystem.InputSettings.shortcutKeysConsumeInput) | Orders overlapping bindings by composite **complexity** (binding-chain depth). This is the default develop behavior when action priority is off. |

Reviewer: ```suggestion
| [Complexity-Based Shortcut Resolution](xref:UnityEngine.InputSystem.InputSettings.shortcutKeysConsumeInput) | Orders overlapping bindings by composite complexity (that is, the binding-chain depth). This is the default develop behavior when the Action Priority option is disabled. |
```

Reviewer: What does "develop behavior" mean?

> | [Action Priority Shortcut Resolution](xref:UnityEngine.InputSystem.InputSettings.shortcutKeysUseActionPriority) | Orders overlapping bindings by each action's [`InputAction.Priority`](xref:UnityEngine.InputSystem.InputAction.Priority). When enabled, it **takes precedence** over complexity-based resolution even if both options are on. |

Reviewer: ```suggestion
| [Action Priority Shortcut Resolution](xref:UnityEngine.InputSystem.InputSettings.shortcutKeysUseActionPriority) | Orders overlapping bindings by each action's [`InputAction.Priority`](xref:UnityEngine.InputSystem.InputAction.Priority). When enabled, this takes precedence over complexity-based resolution, even if both options are enabled. |
```

> Each action has a [`Priority`](xref:UnityEngine.InputSystem.InputAction.Priority) property (range **0**–**65535**, clamped when set). The value applies to all bindings on that action. Serialized priority is always stored on the asset; at runtime it is used only when **Action Priority Shortcut Resolution** is enabled. In that case the **Priority** field is also shown in the [Input Actions Editor](xref:input-system-configuring-input).

Reviewer 2: Might be good to point out the direction of the priority system, e.g. is priority 1 higher or lower than priority 2? That is not obvious.

Author: Done!

> Each action has a [`Priority`](xref:UnityEngine.InputSystem.InputAction.Priority) property (range **0**–**65535**, clamped when set; higher value = higher priority, notified first). The value applies to all bindings on that action. Serialized priority is always stored on the asset; at runtime it is used only when **Action Priority Shortcut Resolution** is enabled. In that case the **Priority** field is also shown in the [Input Actions Editor](xref:input-system-configuring-input).

Reviewer: ```suggestion
Each action has a [`Priority`](xref:UnityEngine.InputSystem.InputAction.Priority) property. The range is from 0 to 65535, and is clamped when set. A higher value means a higher priority, notified first. 

The `Priority` value applies to all bindings on that action. Serialized priority is always stored on the asset; at runtime it's used only when **Action Priority Shortcut Resolution** is enabled. In that case, the **Priority** field is also shown in the [Input Actions Editor](xref:input-system-configuring-input).
```

When action priority resolution is active:

- Higher priority actions are notified before lower-priority actions on the same control.

> - When an action reaches the **Performed** phase, priority **0** does **not** mark the input event as handled, so lower-priority actions in the same overlap group can still respond on that event.

Reviewer 2: My changes in https://github.com/Unity-Technologies/InputSystem/pull/2412 which landed while this has been in development should allow this even when marked handled. Did I miss something in my PR or is something changing on this PR related to this?

> - When an action reaches the **Performed** phase, priority **0** does **not** mark the input event as handled, so lower-priority actions in the same overlap group can still respond on that event.

Reviewer: ```suggestion
- When an action reaches the Performed phase, priority `0 doesn't mark the input event as handled, so lower-priority actions in the same overlap group can still respond on that event.
```

> - Any priority **greater than zero** can mark the event handled and suppress strictly lower-priority actions in the same group for that event.

Reviewer 2: I get a bit nervous about this statement, are you relying on the handled flag to handle priority? What happens then if you have two interactions/bindings like these:

Shift+B - Prio 1
Shift+C - Prio 2

OR 

Shift+B - Prio 2
Shift+C - Prio 1

And the received event sequence is:

First keyboard event: Press shift
Second keyboard event: Both B and C gets pressed in the same atomic event?

Author: Tests should already be covering these cases!

Some info with the help of AI also:

Priority suppression in FireStateChangeNotifications (via signals.ClearBit) only clears monitors within the same overlap group. Groups are assigned per physical control — Shift is in one group, B is in another, C in another. So when B and C arrive in the same atomic event, they're in separate groups and neither suppresses the other. Both Shift+B and Shift+C fire regardless of priority ordering.

Author: Discussed and resolved!

> - Any priority **greater than zero** can mark the event handled and suppress strictly lower-priority actions in the same group for that event.

Reviewer: ```suggestion
- Any priority greater than zero can mark the event as handled and suppress strictly lower-priority actions in the same group for that event.
```

> - Actions with the **same** priority are not suppressed relative to each other; both can perform in the same update if their bindings fire.

Reviewer: ```suggestion
- Actions with the same priority are not suppressed relative to each other; both can perform in the same update if their bindings fire.
```

> Set priority in code:

Reviewer: ```suggestion
To set priority in code, use `.Priority`:
```

```CSharp
fireAction.Priority = 10;
reloadAction.Priority = 5;
```

> Or edit the **Priority** field on an action in the Input Actions Editor when **Action Priority Shortcut Resolution** is enabled.

Reviewer: ```suggestion
You can also edit the **Priority** field on an action in the Input Actions Editor when **Action Priority Shortcut Resolution** is enabled.
```

> For composite shortcuts and complexity ordering, see [Multiple input sequences (such as keyboard shortcuts)](xref:input-system-action-bindings#multiple-input-sequences-such-as-keyboard-shortcuts).

Reviewer: ```suggestion
For information about composite shortcuts and complexity ordering, refer to [Multiple input sequences (such as keyboard shortcuts)](xref:input-system-action-bindings#multiple-input-sequences-such-as-keyboard-shortcuts).
```
